File: platforms/ios/Sources/Feedback/Storage.swift
import Foundation

/**
 Same seam as `KeyValueStore` in core and Android. A protocol rather than a direct
 UserDefaults dependency so the queue and consent logic stay testable with no
 simulator and no UIKit.
 */
public protocol KeyValueStore: AnyObject {
    func get(_ key: String) -> String?
    func set(_ key: String, _ value: String)
    func remove(_ key: String)
    /// Keys carrying the given prefix. The outbox enumerates its entries this way.
    func keys(prefix: String) -> [String]
}

public final class UserDefaultsStore: KeyValueStore {
    private let defaults: UserDefaults

    public init(suiteName: String = "io.markerusa.feedback") {
        // Falls back to .standard when the suite cannot be created (which happens
        // if the name collides with the bundle id) rather than dropping writes.
        self.defaults = UserDefaults(suiteName: suiteName) ?? .standard
    }

    public func get(_ key: String) -> String? {
        defaults.string(forKey: key)
    }

    public func set(_ key: String, _ value: String) {
        defaults.set(value, forKey: key)
    }

    public func remove(_ key: String) {
        defaults.removeObject(forKey: key)
    }

    public func keys(prefix: String) -> [String] {
        defaults.dictionaryRepresentation().keys.filter { $0.hasPrefix(prefix) }
    }
}

/// In-memory store for unit tests.
public final class InMemoryStore: KeyValueStore {
    private var map: [String: String] = [:]

    public init() {}

    public func get(_ key: String) -> String? { map[key] }
    public func set(_ key: String, _ value: String) { map[key] = value }
    public func remove(_ key: String) { map.removeValue(forKey: key) }
    public func keys(prefix: String) -> [String] { map.keys.filter { $0.hasPrefix(prefix) } }
}

File: platforms/android/signals/src/main/kotlin/com/susatest/signals/Storage.kt
package com.susatest.signals

import android.content.Context
import android.content.SharedPreferences

/**
 * Same seam as `KeyValueStore` in core. An interface rather than a direct
 * SharedPreferences dependency so the queue and consent logic stay JVM-testable
 * without an emulator — which is what makes them testable at all on a machine
 * with no Android device attached.
 */
interface KeyValueStore {
    fun get(key: String): String?
    fun set(key: String, value: String)
    fun remove(key: String)
    /** Keys carrying the given prefix. The outbox enumerates its entries this way. */
    fun keys(prefix: String): List<String>
}

internal class SharedPrefsStore(context: Context) : KeyValueStore {
    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences("com.susatest.signals", Context.MODE_PRIVATE)

    override fun get(key: String): String? = prefs.getString(key, null)

    override fun set(key: String, value: String) {
        // commit(), not apply(): the queue's whole purpose is surviving a process
        // death that can happen milliseconds later, and apply() is asynchronous.
        prefs.edit().putString(key, value).commit()
    }

    override fun remove(key: String) {
        prefs.edit().remove(key).commit()
    }

    override fun keys(prefix: String): List<String> =
        prefs.all.keys.filter { it.startsWith(prefix) }
}

/** In-memory store for unit tests. */
class InMemoryStore : KeyValueStore {
    private val map = LinkedHashMap<String, String>()
    override fun get(key: String): String? = map[key]
    override fun set(key: String, value: String) { map[key] = value }
    override fun remove(key: String) { map.remove(key) }
    override fun keys(prefix: String): List<String> = map.keys.filter { it.startsWith(prefix) }
}
